### DISCUSSION TITLE
Buildkite issue

### DISCUSSION THREAD

**Discussion by jagoosw** [CliMA/Oceananigans.jl#3661] [Ideas]
I gather that the CI failures that sometimes happen because all builds put julia in the same place:
https://github.com/CliMA/Oceananigans.jl/blob/eb7c1781bfeea6b4f5d2c5931c999b3d75f78464/.buildkite/pipeline.yml#L16-L17

So if there are multiple PRs happening at the same time they try and install it in the same place, so one build might try and run julia as another is unzipping and then its missing.

Since it is downloaded everytime anyway couldn't each build have its own version e.g.
```bash
      - "wget -N -P $SVERDRUP_HOME https://julialang-s3.julialang.org/bin/linux/x64/$JULIA_MINOR_VERSION/julia-$JULIA_VERSION-linux-x86_64.tar.gz"
      - "tar xf $SVERDRUP_HOME/julia-$JULIA_VERSION-linux-x86_64.tar.gz -C $SVERDRUP_HOME/julia-$BUILDKITE_BUILD_NUMBER"
```
...
```bash
- "$SVERDRUP_HOME/julia-$BUILDKITE_BUILD_NUMBER/julia-$JULIA_VERSION/bin/julia -O0 --color=yes --project -e 'using Pkg; Pkg.instantiate(; verbose=true)'"
```
and then delete them after?

Obviously its more overhead but it might stop the CI issues?

**Comment by glwagner** [comment]
That's a good catch! I think we already use a different _depot_ for each build, but this might be needed too...

I think we should try it. Deleting stuff after the build sometimes fails (so we have to go in and manually delete stuff from time to time) but that's ok, we already have to do that anyways.

I think it will actually be easier to manage all the files if we create a directory for every buildkite build that contains both the julia installation _and_ the depot, and then delete that directory when the build is done. Let me know if that makes sense...

**Reply by jagoosw** [reply]
Yeah that makes sense, I'll make a PR

**Comment by jagoosw** [comment]
Update to this, I tried using the buildkite plugin for OceanBioME and it seems like it is more reliable, but it takes >5 times longer to run the test suite.

I have these two PRs https://github.com/OceanBioME/OceanBioME.jl/pull/190 and testing it with https://github.com/OceanBioME/OceanBioME.jl/pull/196

Another thought I had was that it might be over all more efficient to prevent any test concurrency which I think you could do by making each machine only have one buildkite agent. Given that one when there are concurrent jobs at least one if not both normally fail this would probably be a good intermediate step until we can work out why the plugin is so much slower?

**Reply by glwagner** [reply]
Why would it be so much slower?

**Reply by glwagner** [reply]
I suspect with just 1 agent per machine (rather than 16 as we currently use) the tests would run very slow (16x slower?) but we can test that

**Reply by jagoosw** [reply]
I'm not sure why it's so slow, there are a lot of caching options that I've not played with yet.

I forgot that the Oceananigans tests run different tasks in parallel, I wonder if there is a way to prevent concurrent builds since that's the actual problem?

**Reply by glwagner** [reply]
Why are concurrent builds a problem?

**Reply by glwagner** [reply]
I don't think concurrent builds are the issue. One indication is that we always see race conditions --- not only when we trigger CI on a busy day with lots of other CI.

Different builds are given different julia depots and so should be independent from one another...

It seems like one issue might be that we are not initializing correctly / not initializing enough. We battled this issue when we built the CI, which is why there is an `init` step that instantiates and precompiles:

https://github.com/CliMA/Oceananigans.jl/blob/3c95e7ee6772a2597e206ae35202500be1ef5b32/.buildkite/pipeline.yml#L34-L54

and here:

https://github.com/CliMA/Oceananigans.jl/blob/3c95e7ee6772a2597e206ae35202500be1ef5b32/.buildkite/pipeline.yml#L28-L29

but I think we are missing some stuff there now. Maybe the way the package manager works has changed.

Also possibly a nicer strategy is to run the "init" commands via `runtests` (rather than running an empty test group). 

https://github.com/CliMA/ClimaOcean.jl/blob/4be17784c1f79bf367d9ee922743c83c106d77b5/test/runtests.jl#L7-L10

for some reason that seemed to work better than

**Reply by jagoosw** [reply]
My impression was that some of the failures are coming from if when a build is running, then another starts and redownloads Julia, then the first tries todo something when Julia is only partially downloaded/un zipper?

In OceanBioME (which has an almost identical setup just with far fewer builds occurring) the tests pretty much never fail if only one is running, but almost always do fail if two are running concurrently.

Perhaps just removing the downloading of Julia with every build would solve the issue and we can just manually update it for new releases?

**Reply by glwagner** [reply]
Hmm, but we tried also creating independent julia installs (in addition to independent depots) for different builds and still saw seg faults. There could be multiple causes of the segfaults though. Do you use independent depots in OceanBioME?

**Reply by glwagner** [reply]
Okay, I think I have observed the issue you're referring to @jagoosw, where two different agents conflict with one another over the same julia installation. It seemed to happen here: https://buildkite.com/clima/oceananigans/builds/17319#0191beaa-9b4d-4e69-8120-ae787cee541d/16-31

Notably this caused a failure of the `initialize` step. This arose because sverdrup ran out of space, so when I fixed the problem all 16 agents started simultaneously.

I think the failures we usually see are in subsequent steps though --- not the initialize step. Let me know if that makes sense...
